src/os/bluestore/BlueFS.cc

```diff
@@ -606,6 +615,181 @@ void BlueFS::_stop_alloc()
   }
 }
 
+bool is_all_zeros(const char* p, size_t s)
+{
+  size_t skip = p2nphase(uintptr_t(p), uintptr_t(8));
+  // first check main bulk part
+  if (s > skip) {
+    const uint64_t* gp = (uint64_t*)(p + skip);
+    size_t gs = (s - skip) / sizeof(uint64_t);
+    for (size_t i = 0; i < gs; i++, gp++) {
+      if (*gp != 0) {
+	return false;
+      }
+    }
+  }
+  // now check begin ...
+  for (size_t i = 0; i < skip; i++) {
+    if (p[i] != 0) {
+      return false;
+    }
+  }
+  // ... and tail
+  size_t tail = skip + ((s - skip) & -sizeof(uint64_t));
+  for (size_t i = tail; i < s; i++) {
+    if (p[i] != 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+int BlueFS::read(uint8_t ndev, uint64_t off, uint64_t len,
+		 ceph::buffer::list *pbl, IOContext *ioc, bool buffered)
+{
+  dout(10) << __func__ << " dev " << int(ndev)
+           << ": 0x" << std::hex << off << "~" << len << std::dec
+	   << (buffered ? " buffered" : "")
+	   << dendl;
+  int r;
+  bufferlist bl;
+  r = bdev[ndev]->read(off, len, &bl, ioc, buffered);
+  if (r != 0) {
+    return r;
+  }
+  uint64_t block_size = bdev[ndev]->get_block_size();
+  if (inject_read_zeros) {
+    if (len >= block_size * 2) {
+      derr << __func__ << " injecting error, zeros at "
+	   << int(ndev) << ": 0x" << std::hex << (off + len / 2)
+	   << "~" << (block_size * 2) << std::dec << dendl;
+      //use beginning, replace 8K in the middle with zeros, use tail
+      bufferlist temp;
+      bl.splice(0, len / 2 - block_size, &temp);
+      temp.append(buffer::create(block_size * 2, 0));
+      bl.splice(block_size * 2, len / 2 - block_size, &temp);
+      bl = temp;
+      inject_read_zeros--;
+    }
+  }
+  //make a check if there is a block with all 0
+  uint64_t to_check_len = len;
+  uint64_t skip = p2nphase(off, block_size);
+  if (skip >= to_check_len) {
+    return r;
+  }
+  auto it = bl.begin(skip);
+  to_check_len -= skip;
+  bool all_zeros = false;
+  while (all_zeros == false && to_check_len >= block_size) {
+    // checking 0s step
+    unsigned block_left = block_size;
+    unsigned avail;
+    const char* data;
+    all_zeros = true;
+    while (all_zeros && block_left > 0) {
+      avail = it.get_ptr_and_advance(block_left, &data);
+      block_left -= avail;
+      if (!is_all_zeros(data, avail)) {
```

Comment:
nit: 
all_zeros = is_all_zeros(..);
is enough IMO, no need for "if" and "break".

Reply:
Thanks!